pkg/codegen/json/README.md

```diff
@@ -0,0 +1,98 @@
+# PCL to JSON (experimental)
+
+This directory contains a PCL to JSON code generator which converts a PCL program into a JSON format that represents the Abstract Syntax Tree (AST) of the program. This JSON can be used by external tools to perform analysis or conversion of PCL programs to target languages without needing to build a specialized PCL parser. The JSON format is not intended to be human readable, but rather to be easily parsed by external tools.
+
+## Usage
+Execute `pulumi convert` against a Pulumi YAML program in experimental mode:
+```
+PULUMI_EXPERIMENTAL=1 pulumi convert --language json --out json
+```
+
+Note that Pulumi already type checks the program before converting it to JSON. If the program contains errors, the conversion will fail.
+
+## Format
+
+A PCL program is converted into a JSON object with a field called `nodes` which is an array of the nodes of the program:
+```json
+{
+    "nodes": [...]
+}
+```
+Each `node` can be one of the following:
+
+`Resource` of the following shape:
+```json
+{
+    "type": "Resource",
+    "name": "string",
+    "token": "string",
+    "logicalName": "string",
```

Comment:
I'm not totally sure why we even have logicalName and name? A single resource at the PCL level only really has one name which is the name of it's block. It might also set a name property but that should just get picked up as an input.
I know @AaronFriel added this, but going back to the PR doesn't actually have much context for _why_ we added this (it talks about the difference between logical names and lexical names, but again in PCL I don't see why those aren't the same).

Reply:
`name` and `logicalName` are not the same. I think `name` is the original YAML-compatible name such as `my-bucket` whereas `logicalName` is `myBucket` if I am not mistaken.

Reply:
Both of those are valid PCL names though. I'm not sure what if anything could generate PCL where these differ, and I'm not sure what if anything could do something sensible in code generation when these differ?
I suppose dotnet etc could set variable names based on it, like `var <Name> = new <ResourceType>(<LogicalName>, <Args>)` but I'm not sure if they do? And even if they did each language has it's own constraints around variable names so it has to sanitize and uniqify the names then, it could just work from the logical names just as well.
And again, what would ever generate PCL where these differ? Terraform files only have a single name, YAML files only have a single name. I'm not sure what problem this is solving?

Reply:
Logical names were added to ensure that when converting, we could ensure that the string argument (logical name) was preserved to avoid deletes: https://github.com/pulumi/pulumi-yaml/issues/160

There are as you know three names for a resource, and PCL lacked a means to specify one of them:

```
const lexicalName = new SomeResource(logicalName, {});
const physicalName = lexicalName.id;
```

Reply:
But logical names aren't an arbitrary sequence of bytes... And I'm YAML the lexical name = logical name, and in other languages the generator has to mutate what your calling lexical name to be safe for that language anyway. The whole thing looks suspect to me.

Reply:
```yaml
resources:
  demo-ns:
    type: kubernetes:core/v1:Namespace
```

How do you translate this to Python without deleting?

Reply:
See: https://github.com/pulumi/pulumi/tree/master/pkg/codegen/testing/test/testdata/logical-name-pp and subtests.

Reply:
Logical name absolutely is an arbitrary sequence of bytes. (Or very nearly - I think `:` is a forbidden character?)

Reply:
`deno_ns = Namespace("deno-ns", ...)`
But why should PCL care about these variable naming restrictions in Python?

Reply:
Because PCL is an intermediate language, so how do you express the PCL?

Reply:
See: https://github.com/pulumi/pulumi-yaml/issues/160#issuecomment-1106213639

Reply:
IIRC, `deno-ns` was not a valid HCL identifier because it's parsed as subtraction.

Reply:
> IIRC, deno-ns was not a valid HCL identifier because it's parsed as subtraction.

https://github.com/hashicorp/hcl/blob/main/hclsyntax/spec.md#identifiers

"Identifier = ID_Start (ID_Continue | '-')*;" <- dashes are allowed

> Outputs should have logical names as well because the logical name is "part of the API surface"

I guess, similar to the resource you end up with "LogicalName the string that must be kept as is exactly in the thing sent via GRPC whatever that looks like for your language" and "Name the _suggested_ lexical name for this item if you need a separate lexical name for this item in your langauge".

That really ought to be documented and tested better though, it doesn't look like we have _any_ logicalName tests in pu/pu.

Reply:
We have logical name codegen tests, `logical-names.pp`.

Reply:
Huh missed that when I was looking for this earlier, must of mistyped logical name in my search. 
I'll add a config entry to that as well to show that should have __logicalName.
Also interestingly that test shows that lexicalName isn't even always used, e.g. dotnet's output just uses the logicalName.

Reply:
In fact none of the languages tested there use the lexical name for the output, although I guess they could with code generator changes. E.g. if dotnet used the attribute output style instead of dict return.

Reply:
Yeah, pre-RunAsync style programgen used an attribute:

```cs
    [Output("policyName")]
    public Output<string> PolicyName { get; set; }
```

I felt it was useful to keep lexical & logical names separate, but it does mean we no longer have a test case for the lexical name.

Reply:
> test case for the lexical name appearing in programgen.

Well we do, its just all the current languages don't use the lexical name and that's valid and tested against.